src/components/elements/upload/Uploader.tsx

```diff
@@ -0,0 +1,393 @@
+import { ensureArray } from '@/utils/arrays';
+
+const DEFAULT_MAX_BYTES = 3000000;
+const IMAGE_FILE_TYPES = ['.png', '.jpg', '.jpeg', '.gif'];
+
+const getFileTypesFromAccept = (accept: Accept) => {
+  let arr = sortBy(
+    compact(uniq(flatten(Object.values(accept)))).map((e) =>
+      e.toUpperCase().replace(/\.(.*)/, '$1')
+    )
+  );
+
+  if (arr.length === 2) arr = [arr.join(' or ')];
+  if (arr.length > 2) arr = [...arr.slice(0, -2), arr.slice(-2).join(' or ')];
+
+  return arr.join(', ');
+};
+
+const getReadableSize = (maxSize: number) =>
+  `${(maxSize / 1000000).toFixed(1)}MB`;
+
+export type SingleUploaderProps = {
+  multiple?: false;
+  file?: string | FileFieldsFragment;
+  onChange?: (file: string | FileFieldsFragment | undefined) => void;
+  onUpload?: (
+    upload: DirectUpload | undefined,
+    file: File | undefined
+  ) => any | Promise<any>;
+};
+
+export type MultipleUploaderProps = {
+  multiple: true;
+  files?: (string | FileFieldsFragment)[];
+  onChange?: (files: (string | FileFieldsFragment)[]) => void;
+  onUpload?: (uploads: DirectUpload[], files: File[]) => any | Promise<any>;
+};
+
+export type UploaderProps = {
+  accept?: Accept;
+  image?: boolean;
+  maxSize?: number;
+  id: string;
+  ariaLabel?: string | null;
+} & (SingleUploaderProps | MultipleUploaderProps);
+
+const Uploader = ({
+  id,
+  onChange,
+  onUpload,
+  accept: acceptProp,
+  image: isImage = false,
+  maxSize = DEFAULT_MAX_BYTES,
+  multiple,
+  ariaLabel,
+  ...rest
+}: UploaderProps) => {
+  // The uploader accepts a `files` argument which can contain either:
+  // - a STRING which points at a blob ID of a file that has been uploaded within this session, or
+  // - a FileFieldsFragment record which points at a file record in our database, uploaded during a previous session.
+  // It needs to accept both because this can be a controlled component, so the parent may be keeping track of both previous and current upload state.
+  // `existingFiles` filters the list to only those files that were uploaded some previous time, so we can render them.
+  // The files uploaded during this session, we render this component's internal state, `currentFiles`.
+  const existingFiles: FileFieldsFragment[] = useMemo(() => {
+    // Handles the differing API for a multi vs. single upload component (files vs. file)
+    let filesArr;
+    if (multiple) {
+      const { files } = rest as MultipleUploaderProps;
+      filesArr = ensureArray(files);
+    } else {
+      const { file } = rest as SingleUploaderProps;
+      filesArr = ensureArray(file);
+    }
+    return filesArr.filter(
+      // Filter out files from the input that are just blob IDs.
+      // These should also be reflected in the currentFiles internal state object.
+      (f) => typeof f !== 'string'
+    ) as FileFieldsFragment[]; // Cast to keep typescript happy; now that we've filtered out all the strings, they should all be FileFieldsFragments
```

Comment:
The types can also be differentiated by using a type guard:


```ts
const isFileFieldsFragment = (
  value: string | FileFieldsFragment
): value is FileFieldsFragment =>
  !isNil(value) && typeof value === 'object' && value.__typename === 'File';

```
```suggestion
    // Filter out files from the input that are just blob IDs.
    // These should also be reflected in the currentFiles internal state object.
    return filesArr.filter((f) => isFileFieldsFragment(f));
```

Reply:
Cool, thank you